### PR TITLE
Fixed a few normalize issues

### PR DESCRIPTION
**Pull request by duckki**
Added `normalize` call in `normalize_operation`.

Fixed a bug in NamedFragments::reduce

- normalize selection sets after retain_fragments call

Fixed InlineFragmentSelection::normalize

- Normalize sub-selections with the correct type

Sorry for mixing a few different changes in this PR. Unfortunately, those issues are intertwined in tests. Gathering them in one PR to minimize spurious test changes.

This depends on https://github.com/apollographql/router/pull/5284 in order to pass all tests.
This PR will improve a few tests failing in https://github.com/apollographql/router/pull/5214.

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [X] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [X] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

### CODE REVIEW

File: apollo-federation/src/query_plan/operation.rs [hunk 1]
```diff
@@ -4125,12 +4130,13 @@ impl InlineFragmentSelection {
...
                     );
-
+                    // Return `... on <rebased condition> { __typename @include(if: false) }`
+                    let rebased_casted_type = rebased_fragment.data().casted_type();
                     return Ok(Some(SelectionOrSet::Selection(
                         Selection::from_inline_fragment(
                             rebased_fragment,
                             SelectionSet::from_selection(
-                                parent_type.clone(),
+                                rebased_casted_type,
```
Comment (goto-bus-stop) [inline apollo-federation/src/query_plan/operation.rs:4139]: Hmmm! we should make `Selection::from_inline_fragment` should enforce this. The SelectionSet must always have the fragment's type condition as its parent type (or the fragment's parent type if it does not have a type condition).

Comment (dariuszkuc) [reply apollo-federation/src/query_plan/operation.rs:4139]: I don't think this is necessary true as inline fragment without type condition is still valid i.e. `.... { x y }` is a valid (useless) fragment.

Comment (dariuszkuc) [reply apollo-federation/src/query_plan/operation.rs:4139]: *this was one of the issues i hit with debugging `handle_requires` tests

Comment (goto-bus-stop) [reply apollo-federation/src/query_plan/operation.rs:4139]: Yes, then the selection type must match the fragment's parent type.

Comment (duckki) [reply apollo-federation/src/query_plan/operation.rs:4139]: inline fragment's sub-selection's type must be the same as its casted type. I'll add this assertion.

**Comment by duckki** [comment]
I just added an assertion.
I also reviewed other `from_inline_fragment()` calls. I'll create a separate PR for that later.

**Comment by duckki** [comment]
Oh no. I merged into the wrong branch. I was going to merge into the dev branch. 🤦

**Pull request merged**